src/c_api/normalize.rs

```diff
@@ -0,0 +1,343 @@
+    pub binary: ManuallyDrop<blaze_user_addr_meta_binary>,
+    /// Valid on [`blaze_user_addr_meta_kind::BLAZE_USER_ADDR_KIND_UNKNOWN`].
+    pub unknown: ManuallyDrop<blaze_user_addr_meta_unknown>,
+}
+
+impl Debug for blaze_user_addr_meta_variant {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        f.debug_struct(stringify!(blaze_user_addr_meta_variant))
+            .finish()
+    }
+}
+
+
+/// C ABI compatible version of [`UserAddrMeta`].
+#[repr(C)]
+#[derive(Debug)]
+pub struct blaze_user_addr_meta {
+    /// The variant kind that is present.
+    kind: blaze_user_addr_meta_kind,
+    /// The actual variant with its data.
+    variant: blaze_user_addr_meta_variant,
+}
+
+impl From<UserAddrMeta> for blaze_user_addr_meta {
+    fn from(other: UserAddrMeta) -> Self {
+        match other {
+            UserAddrMeta::Binary(binary) => Self {
+                kind: blaze_user_addr_meta_kind::BLAZE_USER_ADDR_KIND_BINARY,
+                variant: blaze_user_addr_meta_variant {
+                    binary: ManuallyDrop::new(blaze_user_addr_meta_binary::from(binary)),
+                },
+            },
+            UserAddrMeta::Unknown(unknown) => Self {
+                kind: blaze_user_addr_meta_kind::BLAZE_USER_ADDR_KIND_UNKNOWN,
+                variant: blaze_user_addr_meta_variant {
+                    unknown: ManuallyDrop::new(blaze_user_addr_meta_unknown::from(unknown)),
+                },
+            },
+        }
+    }
+}
+
+
+/// An object representing normalized user addresses.
+///
+/// C ABI compatible version of [`NormalizedUserAddrs`].
+#[repr(C)]
+#[derive(Debug)]
+pub struct blaze_normalized_user_addrs {
+    /// The number of [`blaze_user_addr_meta`] objects present in `metas`.
+    pub meta_count: usize,
+    /// An array of `meta_count` objects.
+    pub metas: *mut blaze_user_addr_meta,
+    /// The number of [`blaze_normalized_addr`] objects present in `addrs`.
+    pub addr_count: usize,
+    /// An array of `addr_count` objects.
+    pub addrs: *mut blaze_normalized_addr,
+}
+
+impl From<NormalizedUserAddrs> for blaze_normalized_user_addrs {
+    fn from(other: NormalizedUserAddrs) -> Self {
+        Self {
+            meta_count: other.meta.len(),
+            metas: unsafe {
+                Box::into_raw(
+                    other
+                        .meta
+                        .into_iter()
+                        .map(blaze_user_addr_meta::from)
+                        .collect::<Vec<_>>()
+                        .into_boxed_slice(),
+                )
+                .as_mut()
+                .unwrap()
+                .as_mut_ptr()
+            },
+            addr_count: other.addrs.len(),
+            addrs: unsafe {
+                Box::into_raw(
+                    other
+                        .addrs
+                        .into_iter()
+                        .map(blaze_normalized_addr::from)
+                        .collect::<Vec<_>>()
```

Comment:
will this approach pre-allocate the right size of Vec<> upfront, or it will be gradually resizing it. Asking because we do know the final size of the Vec, so it's more efficient to presize it

Reply:
As discussed in person, yeah, the iterator knows it size and collect will take that into account and presize accordingly in advance.